app/bitcoin/bitcoin.js

```diff
@@ -28,35 +42,60 @@ class BitcoinAddress {
     _hasBitcoinAddress() {
         return this.persistence.hasBitcoinAddress();
     }
+}
 
-    checkBalance() {
-        const address = this.persistence.getPublicKey();
-        return this.serverRequests.getBalance(address);
+
+class TransactionInput {
+    constructor(tx, index, value) {
+        this.tx = tx;
+        this.index = index;
+        this.value = value;
+    }
+}
+
+
+class TransactionOutput {
+    constructor(recipient, amount) {
+        this.recipient = recipient;
+        this.amount = amount;
     }
 }
 
+
+//Instantiate BitcoinTransfer to begin building a list of inputs and outputs
 class BitcoinTransfer {
 
-    //TODO: Allow #buildTransaction() to take a list of sources
-    /**
-     * Sends an amount of bitcoin to the destination wallet
-     * @param sourcetx  - The transaction to use as a source for this transaction
-     * @param sourceIndex   - The index of the source output in its transaction
-     * @param to        - Public key of the recipient's bitcoin wallet
-     * @param from      - Private key of the sender's bitcoin wallet
-     * @param amount    - Amount of bitcoin to buildTransaction (in Satoshi)
-     */
-    static buildTransaction(sourcetx, sourceIndex, from, to, amount) {
-        const transaction = new bitcoin.TransactionBuilder();
+    constructor(myAddress) {
+        this.myAddress = myAddress;
+        this.inputs = [];
+        this.outputs = [];
+        this.currentInputValue = 0;
+        this.PERCENTAGE_CONSTANT = 0.2;
+        this.THRESHOLD = 0.000001;
+    }
+
+    addInput(input) {
+        if (input.constructor !== TransactionInput) {
+            console.log('ERROR : input not of type TransactionInput');
+            return;
+        }
 
-        transaction.addInput(sourcetx, sourceIndex);
-        transaction.addOutput(to, parseInt(amount));
+        //Send dust from previous transaction back to user's address
+        this.outputs.add(TransactionOutput(this.myAddress, this.currentInputValue));
 
-        const privateKey = bitcoin.ECPair.fromWIF(from);
-        transaction.sign(0, privateKey);
+        this.currentInputValue = input.value;
+        this.inputs.add(input);
+    }
+
+    addOutput(address) {
+        const amount = this.currentInputValue - this.PERCENTAGE_CONSTANT;
```

Comment:
Does this just subtract `0.2` from the amount? So if I have 1 bitcoin in my pool, will I donate `0.8` bitcoin to the site?
I think this is supposed to be:
`const amount = this.currentInputValue - (this.currentInputvalue * this.PERCENTAGE_CONSTANT)`